File: 2015/12/python/main.py
from pathlib import Path
import re
import json

def total_sum_1(data:str) -> int:
    pattern = r"-?\d+"
    matches = re.findall(pattern,data)
    total = 0
    for n in matches:
        total += int(n)
    return total

def sum_non_red(data):
    if isinstance(data, int):
        return data
    
    elif isinstance(data, list):
        return sum(sum_non_red(item) for item in data)
    
    elif isinstance(data, dict):
        if "red" in data.values():
            return 0
        return sum(sum_non_red(value) for value in data.values())
    
    # String or other non-numeric values
    return 0

def total_sum_2(data: str) -> int:
    json_data = json.loads(data)
    return sum_non_red(json_data)

def main() -> None:
    data = Path("../input.txt").read_text().strip()
    example = Path("../example.txt").read_text().strip()
    print("=" * 20, " Part 1 ", "=" * 20)
    print(f"sum: {total_sum_1(data)}")
    print("=" * 20, " Part 2 ", "=" * 20)
    print(f"sum: {total_sum_2(data)}")

if __name__ == "__main__":
    main()

File: 2015/12/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func totalSum1(data string) int {
	re := regexp.MustCompile(`-?\d+`)
	matches := re.FindAllString(data, -1)
	sum := 0
	for _, match := range matches {
		if num, err := strconv.Atoi(match); err == nil {
			sum += num
		}
	}
	return sum
}

func sumNonRed(data interface{}) int {
	switch v := data.(type) {
	case float64:
		return int(v)
	case []interface{}:
		sum := 0
		for _, item := range v {
			sum += sumNonRed(item)
		}
		return sum
	case map[string]interface{}:
		for _, value := range v {
			if str, ok := value.(string); ok && str == "red" {
				return 0
			}
		}
		sum := 0
		for _, value := range v {
			sum += sumNonRed(value)
		}
		return sum
	}
	return 0
}

func totalSum2(data string) int {
	var jsonData interface{}
	if err := json.Unmarshal([]byte(data), &jsonData); err != nil {
		return 0
	}
	return sumNonRed(jsonData)
}

func main() {
	fileContent, err := os.ReadFile("../input.txt")
	if err != nil {
		fmt.Println("Error reading file:", err)
		os.Exit(1)
	}

	data := string(fileContent)
	sep := strings.Repeat("=", 20)
	fmt.Println(sep, "Part 1", sep)
	fmt.Println("Sum:", totalSum1(data))
	fmt.Println(sep, "Part 2", sep)
	fmt.Println("Sum:", totalSum2(data))
}
